src/adc/mod.rs

```diff
@@ -0,0 +1,1580 @@
+        self.config.difsel = df;
+
+        self.adc_reg.difsel().modify(|_, w| {
+            for i in 0..19 {
+                w.difsel(i).bit(df.get_channel(i).into());
+            }
+            w
+        });
+    }
+
+    /// Reset the sequence
+    #[inline(always)]
+    pub fn reset_sequence(&mut self) {
+        //The reset state is One conversion selected
+        self.adc_reg
+            .sqr1()
+            .modify(|_, w| unsafe { w.l().bits(config::Sequence::One.into()) });
+    }
+
+    /// Returns the current sequence length. Primarily useful for configuring DMA.
+    #[inline(always)]
+    pub fn sequence_length(&mut self) -> u8 {
+        self.adc_reg.sqr1().read().l().bits() + 1
+    }
+
+    /// Returns the address of the ADC data register. Primarily useful for configuring DMA.
+    #[inline(always)]
+    pub fn data_register_address(&self) -> u32 {
+        self.adc_reg.dr() as *const _ as u32
+    }
+
+    /// Calibrate the adc for <Input Type>
+    #[inline(always)]
+    pub fn calibrate(&mut self, it: config::InputType) {
+        match it {
+            config::InputType::SingleEnded => {
+                self.adc_reg.cr().modify(|_, w| w.adcaldif().clear_bit());
+            }
+            config::InputType::Differential => {
+                self.adc_reg.cr().modify(|_, w| w.adcaldif().set_bit());
+            }
+        }
+
+        self.adc_reg.cr().modify(|_, w| w.adcal().set_bit());
+        while self.adc_reg.cr().read().adcal().bit_is_set() {}
+    }
+
+    /// Calibrate the Adc for all Input Types
+    #[inline(always)]
+    pub fn calibrate_all(&mut self) {
+        self.calibrate(config::InputType::Differential);
+        self.calibrate(config::InputType::SingleEnded);
+    }
+
+    /// Configure a channel for sampling.
+    /// It will make sure the sequence is at least as long as the `sequence` provided.
+    /// # Arguments
+    /// * `channel` - channel to configure
+    /// * `sequence` - where in the sequence to sample the channel. Also called rank in some STM docs/code
+    /// * `sample_time` - how long to sample for. See datasheet and ref manual to work out how long you need\
+    ///   to sample for at a given ADC clock frequency
+    pub fn configure_channel<CHANNEL>(
+        &mut self,
+        _channel: &CHANNEL,
+        sequence: config::Sequence,
+        sample_time: config::SampleTime,
+    ) where
+        CHANNEL: Channel<Ad<ADC>, ID = u8>,
+    {
+        //Check the sequence is long enough
+        self.adc_reg.sqr1().modify(|r, w| unsafe {
+            let prev: config::Sequence = r.l().bits().into();
+            if prev < sequence {
+                w.l().bits(sequence.into())
+            } else {
+                w
+            }
+        });
+
+        let channel = CHANNEL::channel();
+
+        //Set the channel in the right sequence field
+        match sequence {
+            config::Sequence::One => self
+                .adc_reg
+                .sqr1()
```

Comment:
`.sqr(n)` should be possible to use here. Same for `smpr`

Reply:
Thanks!